src/DataCollectors/TraceDataCollector/VanguardCollector/DynamicCoverageDataCollectorImpl.cs

```diff
@@ -0,0 +1,307 @@
+                }
+                catch (Exception ex)
+                {
+                    EqtTrace.Error(
+                        "DynamicCoverageDataCollectorImpl.StartVanguard: Failed to start Vanguard for datacollection context sessionID: {0}, with exception: {1}",
+                        context.SessionId,
+                        ex);
+                    this.logger.LogError(context, ex);
+                    throw;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Stop vanguard
+        /// </summary>
+        /// <param name="context">Context</param>
+        protected void StopVanguard(DataCollectionContext context)
+        {
+            EqtTrace.Info("DynamicCoverageDataCollectorImpl.StopVanguard: Calling Stop Vanguard. datacollection context sessionID: {0}", context.SessionId);
+            if (this.Vanguard != null)
+            {
+                this.Vanguard.Stop();
+
+                if (this.fileHelper.Exists(this.coverageFilePath))
+                {
+                    this.dataSink.SendFileAsync(context, this.coverageFilePath, false);
+                }
+
+                this.Vanguard = null;
+            }
+        }
+
+        protected void OnSendFileCompletedEvent(object sender, AsyncCompletedEventArgs e)
+        {
+             this.CleanupDirectory();
+        }
+
+        /// <summary>
+        /// Generate the file name for coverage file.
+        /// </summary>
+        /// <returns>Returns code coverage file name.</returns>
+        private static string GenerateCoverageFileName()
+        {
+            string GetUserName()
+            {
+                return Environment.GetEnvironmentVariable("USERNAME") ?? Environment.GetEnvironmentVariable("USER");
+            }
+
+            return string.Format(
+                CultureInfo.InvariantCulture,
+                "{0}_{1}_{2}.coverage",
+                GetUserName(),
+                Environment.MachineName,
+                DateTime.Now.ToString("yyyy-MM-dd.HH_mm_ss", CultureInfo.InvariantCulture));
+        }
+
+        private void SetCoverageFileName(XmlElement configurationElement)
+        {
+            XmlElement coverageFileNameElement = configurationElement[CoverageFileSettingName];
+            this.coverageFileName = coverageFileNameElement != null
+                ? coverageFileNameElement.InnerText
+                : GenerateCoverageFileName();
+        }
+
+        private void PrepareVanguardProcess(XmlElement configurationElement)
+        {
+            EqtTrace.Info("DynamicCoverageDataCollectorImpl.PrepareVanguardProcess: Preparing Vanguard process.");
+
+            XmlElement config = configurationElement[ConfigCodeCoverageElementName];
+            string configurationFileName = Path.Combine(this.sessionDirectory, VanguardConfigFileName);
+
+            this.fileHelper.WriteAllText(configurationFileName, config.OuterXml);
```

Comment:
Potential exception point, please log in Trace, & use IMessageLogger to send back error